File: src/de/rub/bi/inf/openbimrl/functions/filter/FilterInvert.java
package de.rub.bi.inf.openbimrl.functions.filter;

import de.rub.bi.inf.openbimrl.NodeProxy;
import de.rub.bi.inf.openbimrl.engine.ifc.IIFCModel;
import de.rub.bi.inf.openbimrl.functions.AbstractFunction;

import java.util.ArrayList;
import java.util.LinkedHashMap;

/**
 * Inverts the result of a filter. Basically, performing a boolean flip on all values of a list.
 *
 * @author Marcel Stepien
 */
public class FilterInvert extends AbstractFunction {

    public FilterInvert(NodeProxy nodeProxy) {
        super(nodeProxy);
    }

    @Override
    public void execute(IIFCModel ifcModel) {

        Object input0 = getInput(0);

        LinkedHashMap<Object, ArrayList<Boolean>> filterMap = new LinkedHashMap<Object, ArrayList<Boolean>>();

        if (input0 instanceof ArrayList) {
            ArrayList<Boolean> mask = (ArrayList<Boolean>) input0;
            filterMap.put("0", (ArrayList<Boolean>) mask);
        }

        if (input0 instanceof LinkedHashMap) {
            filterMap = (LinkedHashMap<Object, ArrayList<Boolean>>) input0;
        }

        if (filterMap == null)
            return;

        LinkedHashMap<Object, ArrayList<?>> resultValues = new LinkedHashMap<Object, ArrayList<?>>();

        for (Object maskKey : filterMap.keySet()) {
            ArrayList<?> mask = filterMap.get(maskKey);

            ArrayList filteredList = new ArrayList();
            for (int i = 0; i < mask.size(); i++) {
                boolean flag = mask.get(i) != null ? Boolean.parseBoolean(mask.get(i).toString()) : false;
                filteredList.add(!flag);
            }

            resultValues.put(maskKey, filteredList);
        }

        if (resultValues.keySet().size() == 1) {
            Object firstKey = resultValues.keySet().toArray()[0];
            setResult(0, resultValues.get(firstKey));
        } else {
            setResult(0, resultValues);
        }
    }

}

File: src/main/jvm/de/rub/bi/inf/openbimrl/functions/filter/FilterInvert.kt
package de.rub.bi.inf.openbimrl.functions.filter

import de.rub.bi.inf.openbimrl.NodeProxy
import de.rub.bi.inf.openbimrl.functions.AbstractFunction

/**
 * Inverts the result of a filter. Basically, performing a boolean flip on all values of a list.
 *
 * @author Marcel Stepien
 */
class FilterInvert(nodeProxy: NodeProxy) : AbstractFunction(nodeProxy) {
    override fun execute() {
        val input0 = getInput<Any>(0) ?: return

        var filterMap = LinkedHashMap<Any, List<Boolean>>()

        if (input0 is ArrayList<*>) {
            val mask = input0.filterIsInstance<Boolean>()
            filterMap["0"] = mask
        }

        if (input0 is LinkedHashMap<*, *>) {
            // dangerous cast!
            filterMap = input0 as LinkedHashMap<Any, List<Boolean>>
        }

        val resultValues = LinkedHashMap<Any, ArrayList<*>>()

        for (maskKey in filterMap.keys) {
            val mask = filterMap[maskKey] ?: continue

            val filteredList = ArrayList<Any?>()
            for (i in mask.indices) {
                val flag = mask[i]
                filteredList.add(!flag)
            }

            resultValues[maskKey] = filteredList
        }

        if (resultValues.keys.size == 1) {
            val firstKey = resultValues.keys.toTypedArray()[0]
            setResult(0, resultValues[firstKey])
        } else {
            setResult(0, resultValues)
        }
    }
}
